In [35]:
import requests
import os
from dotenv import load_dotenv, find_dotenv
import pandas as pd

In [36]:
load_dotenv(find_dotenv())

True

In [37]:
buildings_token = os.getenv("BUILDINGS")
hca_token = os.getenv("HEAT_COST_ALLOCATORS")
rooms_token = os.getenv("ROOMS")
units_token = os.getenv("UNITS")
hca_details_token = os.getenv("HEAT_COST_ALLOCATOR_DETAILS")
room_details_token = os.getenv("ROOM_DETAILS")

In [38]:
baseurl = "https://edc.e-b-z.de/public"

In [48]:
def get_metadata( token: str, baseurl: str = 'https://edc.e-b-z.de/public'):
    
    resp = requests.get(
        f"{baseurl}",
        headers={'Content-Type': 'application/json',"Authorization": token},
    )
    if 200 <= resp.status_code < 300:
        resp_data = resp.json()
        return resp_data
    else:
        print(f"Error: {resp.status_code} - {resp.text}")
        return None

In [63]:
def parse_json_csv(data, filename='metadata.csv'):
    df = pd.json_normalize(data)
    df.to_csv(filename, index=False)
    return df

In [49]:
building_meta_data = get_metadata(token=buildings_token)

In [51]:
rooms_meta_data = get_metadata(token=rooms_token)

In [61]:
units_meta_data = get_metadata(token=units_token)

In [74]:
building_meta_df = parse_json_csv(building_meta_data, filename='building_metadata.csv')
rooms_meta_df = parse_json_csv(rooms_meta_data, filename='rooms_metadata.csv')
units_meta_df = parse_json_csv(units_meta_data, filename='units_metadata.csv')

In [65]:
units_meta_df

area  building_id  floor  floor_relative    site  site_relative  unit_id
0    60.0           57    1.0             1.0   Links            0.0      357
1    60.0           57    1.0             1.0  Rechts            1.0      358
2    60.0           57    2.0             2.0   Links            0.0      359
3    60.0           57    2.0             2.0  Rechts            1.0      360
4    60.0           57    0.0             0.0   Links            0.0      361
..    ...          ...    ...             ...     ...            ...      ...
347  56.0           13    2.0             2.0    None            NaN     1105
348  55.0           13    2.0             2.0    None            NaN      598
349  61.0           46    2.0             2.0    None            NaN      619
350  68.0           46    3.0             2.0    None            NaN     1075
351   NaN           46    2.0             1.0    None            NaN      620

[352 rows x 7 columns]

In [67]:
building_meta_df

area_building_envelope  area_heat  area_roof  area_underground  \
0                    735.0     754.00      320.0             320.0   
1                    787.5     605.00      260.0             260.0   
2                    489.6        NaN      480.0             480.0   
3                    787.5     615.00      260.0             260.0   
4                      NaN        NaN        NaN               NaN   
5                    652.5     545.00      242.0             242.0   
6                    652.5     547.00      242.0             242.0   
7                      NaN     680.00        NaN               NaN   
8                    751.5     940.00      720.0             720.0   
9                    751.5     940.00      720.0             720.0   
10                   642.0     535.00      243.0             243.0   
11                   642.0     456.00      243.0             243.0   
12                   690.0     522.00      270.0             270.0   
13                     NaN        NaN        NaN               NaN   
14                  1080.0     585.30      380.0             380.0   
15                   652.5     545.00      242.0             242.0   
16                   652.5     547.00      242.0             242.0   
17                   937.5     811.00      396.0             396.0   
18                     NaN     735.00        NaN               NaN   
19                   489.6     604.44      480.0             480.0   
20                     NaN     735.00        NaN               NaN   
21                     NaN     735.00        NaN               NaN   
22                   489.6     602.04      480.0             480.0   
23                     NaN        NaN        NaN               NaN   
24                   474.0     556.00      380.0             380.0   
25                     NaN        NaN        NaN               NaN   
26                   550.0     408.00      160.0             240.0   
27                   429.0        NaN      300.0             300.0   
28                   620.0     616.00      102.0             340.0   
29                   620.0     606.00      102.0             340.0   
30                   572.0     539.00       93.0             310.0   

    area_window_door  building_height  building_id  building_length  \
0              105.0             10.0           58             32.0   
1               76.5             12.0           26             26.0   
2              134.4             12.0           57             12.0   
3               76.5             12.0           52             26.0   
4                NaN              NaN           17              NaN   
5              139.5             12.0            2             22.0   
6              139.5             12.0           45             22.0   
7                NaN              NaN           16              NaN   
8              256.5              8.0           47             48.0   
9              256.5              8.0           50             48.0   
10              78.0             10.0           28             27.0   
11              78.0             10.0           13             27.0   
12              90.0             10.0           46             30.0   
13               NaN              NaN           39              NaN   
14             720.0             12.5           14             15.0   
15             139.5             12.0           53             22.0   
16             139.5             12.0           18             22.0   
17             190.5             12.0           73             36.0   
18               NaN              NaN            7              NaN   
19             134.4             12.0           66             12.0   
20               NaN              NaN           38              NaN   
21               NaN              NaN           74              NaN   
22             134.4             12.0            4             12.0   
23               NaN              NaN           20         

In [78]:
rooms_meta_df_extended = rooms_meta_df.join(units_meta_df.set_index('unit_id'), on='unit_id').reset_index()
rooms_meta_df_extended = rooms_meta_df_extended.drop(columns=['area','floor_relative', 'floor','site', 'site_relative'])

In [80]:
rooms_meta_df_extended.drop(['index'], axis=1, inplace=True)

In [84]:
hca_meta_df = get_metadata(token=hca_token)
hca_meta_df = parse_json_csv(hca_meta_df, filename='hca_metadata.csv')

In [85]:
hca_meta_df

active date_installation date_removal  heat_cost_allocator_id   kcl  \
0       True        2012-06-01         None                     666  1.75   
1       True        2012-06-01         None                     333  1.75   
2       True        2012-06-01         None                     336  1.75   
3       True        2012-06-01         None                    2155  1.75   
4       True        2012-06-01         None                     669  1.75   
...      ...               ...          ...                     ...   ...   
4102   False        2004-03-16   2013-09-23                   12650   NaN   
4103   False        2004-03-16   2013-09-23                   12645   NaN   
4104   False        2004-03-16   2013-09-23                   12643   NaN   
4105   False        2004-03-16   2013-09-23                   12656   NaN   
4106    True              None         None                     827   NaN   

       kcw      qs  room_id  
0     1.16  1345.0     1013  
1     1.16   960.0     1018  
2     1.14  2353.0     1021  
3     1.16  1514.0     1446  
4     1.14  1741.0     1563  
...    ...     ...      ...  
4102   NaN     NaN      278  
4103   NaN     NaN      279  
4104   NaN     NaN      276  
4105   NaN     NaN      280  
4106   NaN     NaN      397  

[4107 rows x 8 columns]

In [83]:
rooms_meta_df_extended.to_csv('rooms_metadata.csv', index=False)

In [70]:
room_id = 939
#curl --location 'https://edc.e-b-z.de/public/1000/temperatures' --header 'Content-Type: application/json' --header 'Authorization: <token>'
resp = requests.get(
    f"{baseurl}/{room_id}/temperatures",
    headers={'Content-Type': 'application/json',"Authorization": room_details_token},
    params={'page': 1}
)
if 200 <= resp.status_code < 300:
    resp_data = resp.json()
resp.status_code

200

In [71]:
resp_data

{'data': [{'room_id': 939,
   'temperature': 19.31,
   'ts': '2017-01-20T01:14:12+00:00'},
  {'room_id': 939, 'temperature': 19.28, 'ts': '2017-01-20T01:19:58+00:00'},
  {'room_id': 939, 'temperature': 19.26, 'ts': '2017-01-20T01:21:50+00:00'},
  {'room_id': 939, 'temperature': 19.23, 'ts': '2017-01-20T01:33:11+00:00'},
  {'room_id': 939, 'temperature': 19.23, 'ts': '2017-01-20T01:36:59+00:00'},
  {'room_id': 939, 'temperature': 19.21, 'ts': '2017-01-20T01:40:49+00:00'},
  {'room_id': 939, 'temperature': 19.16, 'ts': '2017-01-20T01:55:57+00:00'},
  {'room_id': 939, 'temperature': 19.15, 'ts': '2017-01-20T01:59:48+00:00'},
  {'room_id': 939, 'temperature': 19.13, 'ts': '2017-01-20T02:01:40+00:00'},
  {'room_id': 939, 'temperature': 19.12, 'ts': '2017-01-20T02:07:25+00:00'},
  {'room_id': 939, 'temperature': 19.11, 'ts': '2017-01-20T02:09:17+00:00'},
  {'room_id': 939, 'temperature': 19.1, 'ts': '2017-01-20T02:16:54+00:00'},
  {'room_id': 939, 'temperature': 19.09, 'ts': '2017-01-20T02:1

In [132]:
df = pd.json_normalize(resp_data['data'])
df

room_id  temperature                         ts
0       1000        19.28  2023-02-13T08:16:30+00:00
1       1000        19.24  2023-02-13T08:18:28+00:00
2       1000        19.16  2023-02-13T08:25:56+00:00
3       1000        19.12  2023-02-13T08:33:28+00:00
4       1000        19.11  2023-02-13T08:35:21+00:00
..       ...          ...                        ...
995     1000        17.46  2023-02-17T09:37:38+00:00
996     1000        17.42  2023-02-17T09:48:55+00:00
997     1000        17.41  2023-02-17T09:50:45+00:00
998     1000        17.40  2023-02-17T09:52:38+00:00
999     1000        17.41  2023-02-17T09:58:21+00:00

[1000 rows x 3 columns]

In [106]:
rooms_meta_df_extended

room_id type  unit_id  building_id
0         939   KI      357           57
1         941    B      357           57
2         940    K      357           57
3         943    S      357           57
4         942    W      357           57
...       ...  ...      ...          ...
1840    10166   WK      297           46
1841      836    B      297           46
1842      835    K      297           46
1843      838    S      297           46
1844      837    W      297           46

[1845 rows x 4 columns]

In [72]:
hca_id = 13794
#curl --location 'https://edc.e-b-z.de/public/1000/temperatures' --header 'Content-Type: application/json' --header 'Authorization: <token>'
resp = requests.get(
    f"{baseurl}/{hca_id}/temperatures",
    headers={'Content-Type': 'application/json',"Authorization": hca_details_token},
    params={'page': 2}
)
if 200 <= resp.status_code < 300:
    resp_data = resp.json()
resp.status_code

200

In [73]:
resp_data

[]

In [117]:
hca_df = pd.json_normalize(resp_data)

In [119]:
hca_df.to_csv('hca_temp_test.csv', index=False)

In [112]:
hca_meta_df_extended = hca_meta_df.join(rooms_meta_df_extended.set_index('room_id'), on='room_id').reset_index()

In [114]:
hca_meta_df_extended[hca_meta_df_extended['room_id'] == room_id]

index  active date_installation date_removal  heat_cost_allocator_id  \
156    156    True        2014-03-11         None                    2464   
424    424   False        2004-03-16   2014-03-10                   12121   

      kcl   kcw      qs  room_id type  unit_id  building_id  
156  1.75  1.14  4357.0      948    W      358           57  
424   NaN   NaN     NaN      948    W      358           57

In [109]:
df = pd.json_normalize(resp_data['data'])
df.to_csv('room_temp_test.csv', index=False)

In [108]:
room_id = 948
rooms_meta_df_extended[rooms_meta_df_extended['room_id'] == room_id]

room_id type  unit_id  building_id
9      948    W      358           57

In [42]:

baseurl = "https://edc.e-b-z.de/public"
building_id = 58
resp = requests.get(
    f"{baseurl}",
    headers={'Content-Type': 'application/json',"Authorization": hca_token},
)
if 200 <= resp.status_code < 300:
    resp_data = resp.json()
resp_data

[{'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 666,
  'kcl': 1.75,
  'kcw': 1.16,
  'qs': 1345.0,
  'room_id': 1013},
 {'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 333,
  'kcl': 1.75,
  'kcw': 1.16,
  'qs': 960.0,
  'room_id': 1018},
 {'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 336,
  'kcl': 1.75,
  'kcw': 1.14,
  'qs': 2353.0,
  'room_id': 1021},
 {'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 2155,
  'kcl': 1.75,
  'kcw': 1.16,
  'qs': 1514.0,
  'room_id': 1446},
 {'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 669,
  'kcl': 1.75,
  'kcw': 1.14,
  'qs': 1741.0,
  'room_id': 1563},
 {'active': True,
  'date_installation': '2012-06-01',
  'date_removal': None,
  'heat_cost_allocator_id': 332

In [75]:

baseurl = "https://edc.e-b-z.de/public"
hca_id = 336
resp = requests.get(
    f"{baseurl}/{hca_id}/units",
    headers={'Content-Type': 'application/json',"Authorization": hca_details_token},
    params={'from': '2017-01-23', 'to': '2017-01-24', 'per_page': 1000, 'page': 1}
)
if 200 <= resp.status_code < 300:
    resp_data = resp.json()
resp_data

[{'heat_cost_allocator_id': 336,
  'ts': '2017-01-23T22:59:59+00:00',
  'units': 118.0}]

In [76]:
dd = pd.json_normalize(resp_data)
dd

heat_cost_allocator_id                         ts  units
0                     336  2017-01-23T22:59:59+00:00  118.0

In [1]:
import pandas as pd

build_df = pd.read_csv('./metadata/building_metadata.csv')
rooms_df = pd.read_csv('./metadata/rooms_metadata.csv')
hca_df = pd.read_csv('./metadata/hca_metadata.csv')

In [13]:
rooms_df['building_id'].unique()

array([57, 58, 26, 17, 52,  4, 45, 10,  2, 14, 16, 18, 20, 50, 21, 38,  7,
       53, 66, 73, 28, 24, 47, 39, 13, 23, 74, 46])

In [4]:
build_57 = rooms_df[rooms_df['building_id'] == 57]

In [28]:
len(build_57['room_id'].unique())

120

In [11]:
hca_df_57 = hca_df[hca_df['room_id'].isin(build_57['room_id'])]
hca_df_57

active date_installation date_removal  heat_cost_allocator_id   kcl  \
133    True        2014-03-11          NaN                    2497  1.75   
134    True        2014-03-11          NaN                    2508  1.75   
135    True        2014-03-11          NaN                    2509  1.84   
136    True        2014-03-11          NaN                    2495  1.75   
137    True        2014-03-11          NaN                    2494  1.75   
..      ...               ...          ...                     ...   ...   
554   False        2004-03-16   2014-03-10                   12068   NaN   
562   False        2004-03-16   2014-03-10                   12108   NaN   
564   False        2014-03-11   2016-03-14                   30616  1.68   
565   False        2014-03-11   2017-05-02                   32695  1.84   
566   False        2014-03-11   2016-02-17                   32713  1.75   

      kcw      qs  room_id  
133  1.14  1532.0      982  
134  1.14  1532.0      995  
135  1.11  5636.0      996  
136  1.14  1676.0      975  
137  1.14  3064.0      979  
..    ...     ...      ...  
554   NaN     NaN      964  
562   NaN     NaN      947  
564  1.19  1456.0    14201  
565  1.11  1172.0    15512  
566  1.14  1556.0    15527  

[141 rows x 8 columns]

In [26]:
len(hca_df_57[hca_df_57['active']].heat_cost_allocator_id.unique())

66

In [18]:
hca_temp_test = pd.read_csv('hca_temp_test.csv')
room_temp_test = pd.read_csv('room_temp_test.csv')

In [51]:
hca_temp_test.heat_cost_allocator_id.unique()

array([ 2462,  2460,  2458,  2461,  2459,  2467,  2468,  2465,  2463,
        2466,  2464,  2473,  2471,  2469,  2472,  2470,  2478,  2479,
        2476,  2474,  2477,  2475,  2451,  2449,  2447,  2450,  2448,
        2456,  2457,  2454,  2452,  2455,  2453,  2495,  2496, 19807,
        2491,  2494,  2492,  2501,  2499,  2497,  2500,  2498,  2506,
        2507,  2504,  2502,  2505,  2503,  2512,  2510,  2508,  2511,
        2509,  2484,  2485,  2482,  2480,  2483,  2481,  2490,  2488,
        2486,  2489,  2487])

In [46]:
room_temp_test.room_id.unique()

array([ 939,  941,  940,  943,  942,  945,  947,  946,  949,  948,  951,
        953,  952,  955,  954,  957,  959,  958,  961,  960,  963,  965,
        964,  967,  966,  969,  971,  970,  973,  972,  975,  977,  976,
        979,  978,  981,  983,  982,  985,  989,  988,  991,  990,  993,
        994,  995,  997,  996,  999, 1001, 1000, 1003, 1002, 1007, 1006,
       1009, 1008])

In [31]:
rooms_missing = build_57[~build_57['room_id'].isin(room_temp_test['room_id'].unique())]

In [50]:
rooms_missing['room_id'].unique()

array([  984,   987,  1005, 14179, 14180, 14181, 14182, 14183, 14184,
       14185, 14186, 14187, 14188, 14189, 14190, 14191, 14192, 14193,
       14194, 14195, 14196, 14197, 14198, 14199, 14200, 14201, 14202,
       14203, 14204, 14205, 14206, 14207, 14208, 15500, 15501, 15502,
       15503, 15504, 15505, 15506, 15507, 15508, 15509, 15510, 15511,
       15512, 15513, 15514, 15515, 15516, 15517, 15518, 15519, 15520,
       15521, 15522, 15523, 15524, 15525, 15526, 15527, 15528, 15529])

### These rooms for building 57 return [] when we fetch the data
[  984,   987,  1005, 14179, 14180, 14181, 14182, 14183, 14184,
       14185, 14186, 14187, 14188, 14189, 14190, 14191, 14192, 14193,
       14194, 14195, 14196, 14197, 14198, 14199, 14200, 14201, 14202,
       14203, 14204, 14205, 14206, 14207, 14208, 15500, 15501, 15502,
       15503, 15504, 15505, 15506, 15507, 15508, 15509, 15510, 15511,
       15512, 15513, 15514, 15515, 15516, 15517, 15518, 15519, 15520,
       15521, 15522, 15523, 15524, 15525, 15526, 15527, 15528, 15529]

### And These rooms actually return data when we fetch:
array([ 939,  941,  940,  943,  942,  945,  947,  946,  949,  948,  951,
        953,  952,  955,  954,  957,  959,  958,  961,  960,  963,  965,
        964,  967,  966,  969,  971,  970,  973,  972,  975,  977,  976,
        979,  978,  981,  983,  982,  985,  989,  988,  991,  990,  993,
        994,  995,  997,  996,  999, 1001, 1000, 1003, 1002, 1007, 1006,
       1009, 1008])

In [52]:
#rooms_missing = build_57[~build_57['room_id'].isin(room_temp_test['room_id'].unique())]
hca_missing = hca_df_57[~hca_df_57['heat_cost_allocator_id'].isin(hca_temp_test['heat_cost_allocator_id'].unique())]

In [60]:
hca_missing['active'].unique()

array([False])

In [54]:
hca_missing['heat_cost_allocator_id'].unique()

array([12080, 13794, 12141, 12154, 12198, 12100, 12125, 12206, 12087,
       12119, 12212, 13774, 13770, 13772, 12083, 12123, 12193, 12200,
       12089, 13781, 12138, 12217, 12143, 12065, 13790, 12098, 12196,
       12096, 12061, 12056, 12151, 12135, 13786, 12073, 12189, 12159,
       12112, 13776, 12164, 12126, 12121, 13765, 12133, 12203, 12175,
       12131, 12093, 12171, 12219, 12179, 12191, 12058, 12183, 12221,
       12059, 12085,  2493, 12185, 12091, 13788, 12167, 12214, 12104,
       12110, 12147, 13778, 12078, 12162, 12157, 12224, 12068, 12108,
       30616, 32695, 32713])

### Missing HCAs for building 57
array([12080, 13794, 12141, 12154, 12198, 12100, 12125, 12206, 12087,
       12119, 12212, 13774, 13770, 13772, 12083, 12123, 12193, 12200,
       12089, 13781, 12138, 12217, 12143, 12065, 13790, 12098, 12196,
       12096, 12061, 12056, 12151, 12135, 13786, 12073, 12189, 12159,
       12112, 13776, 12164, 12126, 12121, 13765, 12133, 12203, 12175,
       12131, 12093, 12171, 12219, 12179, 12191, 12058, 12183, 12221,
       12059, 12085,  2493, 12185, 12091, 13788, 12167, 12214, 12104,
       12110, 12147, 13778, 12078, 12162, 12157, 12224, 12068, 12108,
       30616, 32695, 32713])

In [63]:
room_id = 984
hca_metadata = pd.read_csv("metadata/hca_metadata.csv")
room_hcas = hca_metadata[(hca_metadata['room_id'] == room_id) & (hca_metadata['active']==True)]

In [64]:
room_hcas

active date_installation date_removal  heat_cost_allocator_id   kcl  \
145    True        2014-03-11          NaN                    2498  1.75   

      kcw      qs  room_id  
145  1.14  4357.0      984